### PR TITLE
TST: require matplotlib binaries in devdeps so that nightlies take priority over pre-releases without wheels

### LINKED ISSUE #15283: TST: devdeps failing with numpy-dev failures from mpl
Thanks to @neutrinoceros for the notice. We will just have to wait it out. Original post on Slack as follows.

<hr/>

In case I can save anyone else from chasing their tail today: there are a couple failures in future-compatibility tests today that are due to matplotlib using newly deprecated API in future numpy. The problem was already fixed upstream 2 weeks ago, it's just that pip sees matplotlib `3.8.0rc1` (which doesn't have the patch) as more "recent" than the nightly wheel, so I don't think there's anything we can do but wait for a new release, or nightlies with a version number starting with `"3.9"`

For reference:

* https://github.com/numpy/numpy/pull/24445
* https://github.com/matplotlib/matplotlib/pull/26566

### PR DESCRIPTION
**Pull request by neutrinoceros**
This tentatively fix #15283

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by neutrinoceros** [comment]
I'm not well versed in tox; the only way I found to pass force binaries was through a constraint file, but it works. Remaining failures are out of scope for this PR and have to do with NEP 18 function being deprecated in NumPy 2.0.

File: tox.ini [hunk 1]
```diff
@@ -109,7 +109,7 @@ extras =
     alldeps: test_all
 
 commands =
-    mpldev: pip install -U --pre -i https://pypi.anaconda.org/scientific-python-nightly-wheels/simple matplotlib
+    mpldev: pip install -U --pre -i https://pypi.anaconda.org/scientific-python-nightly-wheels/simple matplotlib --only-binary matplotlib
```
Comment (pllim) [inline tox.ini:112]: Maybe @ksunden from matplotlib has a better idea? 👀

**Comment by ksunden** [comment]
I'm confused why that would fix it, as we release wheels for our RC.

We don't typically merge the release branch (with its tag) back into main until the final release, and that is what triggers main to bump versions (which then is the wheels we upload to nightlies).

I think what you might actually want is `--force-reinstall` added to this line as I think it is actually prior to this line that it is getting the rc from pypi as `-i` is the form that should only look in the one index.

**Comment by neutrinoceros** [comment]
> I'm confused why that would fix it, as we release wheels for our RC.

Indeed, but `3.8.0rc1` doesn't have `cp312` wheels. My patch wouldn't work (but also would n'y need to, at first) if there was a second RC with wheels for Python 3.12

I agree that forcing reinstall would be more robust, I'm just not sure how to configure that as I'm new to tox.

**Comment by pllim** [comment]
This RC period is how long? I am not sure if this effort is worth all the time already put into it if the problem is going away on its own within 2 weeks or so.

Though I do wonder why mpldev directive differs from devdeps. @astrofrog or @saimn , do you remember why?

**Comment by astrofrog** [comment]
The mpldev config was just for the image tests so we could test with just Matplotlib dev and not also have Numpy dev etc but to be honest I don't think that distinction is really needed anymore and we could probably switch the image test to use devdeps too.

I also don't think this PR is really needed - it's not the end of the world if for a couple of weeks we test against RC instead of true developer versions - it's not ideal but also not critical.

**Comment by pllim** [comment]
@astrofrog , no, I mean why the mpldev install command is different from the matplotlib dev install command in devdeps.

**Comment by astrofrog** [comment]
@pllim 'historical reasons' :smile: - feel free to change

**Comment by neutrinoceros** [comment]
I haven't followed closely but from what I can tell problems appear to be fixed on main ? any way, feel free to close this !

**Comment by pllim** [comment]
Well, the problem didn't really get solved but now the CI is pulling both matplotlib rc *and* numpy 1.26 rc (so we're not plagued by numpy 2.0.dev for now).

**Comment by neutrinoceros** [comment]
gotcha. Then I suppose it doesn't hurt either to keep this one around in case you want to merge it after all, and if it's still not merged when matplotlib 3.8 comes out there'll be no cost to closing it.

**Comment by neutrinoceros** [comment]
oooooh, spooky 👻

**Pull request closed without merging**